sagemaker-neo-notebooks/edge/cpp-integration/tutorial.cc

```diff
@@ -0,0 +1,622 @@
+  const Aws::String aws_object_name(model_name.c_str(), model_name.size());
+
+  // download s3 object
+  Aws::S3::S3Client s3_client = getS3Client();
+
+  Aws::S3::Model::GetObjectRequest object_request;
+  object_request.SetBucket(aws_bucket_name);
+  object_request.SetKey(aws_object_name);
+
+  auto get_object_outcome = s3_client.GetObject(object_request);
+  if (!get_object_outcome.IsSuccess()) {
+    auto error = get_object_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("GetModel error: " + error_str);
+  } else {
+    auto &model_file = get_object_outcome.GetResultWithOwnership().GetBody();
+
+    // download the sample file
+    const char *output_filename = filename.c_str();
+    std::ofstream output_file(output_filename, std::ios::binary);
+    output_file << model_file.rdbuf();
+  }
+}
+
+void createBucket(std::string &bucket_name) {
+  // create bucket
+  const Aws::String s3_bucket_name(bucket_name.c_str(), bucket_name.size());
+  Aws::S3::Model::CreateBucketRequest request;
+  request.SetBucket(s3_bucket_name);
+
+  // for demo purpose: we set that to us-west-2
+  const Aws::S3::Model::BucketLocationConstraint &region =
+      Aws::S3::Model::BucketLocationConstraint::us_west_2;
+  Aws::S3::S3Client s3_client = getS3Client();
+  Aws::S3::Model::CreateBucketConfiguration bucket_config;
+  bucket_config.SetLocationConstraint(region);
+  request.SetCreateBucketConfiguration(bucket_config);
+
+  // making request call
+  auto outcome = s3_client.CreateBucket(request);
+  if (!outcome.IsSuccess()) {
+    auto err = outcome.GetError();
+    std::string error_str = getErrorMessage(err);
+    throw std::runtime_error("CreateBucket error: " + error_str);
+  }
+}
+
+void uploadModel(std::string &bucket_name, std::string &model_name,
+                 std::string &filename) {
+  Aws::String s3_bucket_name(bucket_name.c_str(), bucket_name.size());
+  Aws::String s3_aws_object_name(model_name.c_str(), model_name.size());
+  Aws::S3::S3Client s3_client = getS3Client();
+
+  // upload model to s3 bucket
+  Aws::S3::Model::PutObjectRequest object_request;
+  object_request.SetBucket(s3_bucket_name);
+  object_request.SetKey(s3_aws_object_name);
+
+  // set binary to body
+  const std::shared_ptr<Aws::IOStream> input_data =
+      Aws::MakeShared<Aws::FStream>("SampleAllocationTag", filename.c_str(),
+                                    std::ios_base::in | std::ios_base::binary);
+  object_request.SetBody(input_data);
+
+  // put model to the s3 bucket
+  auto put_object_outcome = s3_client.PutObject(object_request);
+  if (!put_object_outcome.IsSuccess()) {
+    auto error = put_object_outcome.GetError();
+    std::string error_str = getErrorMessage(error);
+    throw std::runtime_error("UploadModel error: " + error_str);
+  }
+}
+
+bool checkBucketExist(std::string &bucket_name) {
+  Aws::String s3_bucket_name(bucket_name.c_str(), bucket_name.size());
```

Comment:
I think they have ctor from `std::string` https://github.com/aws/aws-sdk-cpp/blob/master/aws-cpp-sdk-core/include/aws/core/utils/memory/stl/AWSString.h#L83

Reply:
believe this is for std::basic_string; hence this is not going to work.